src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListFilter.cs

```diff
@@ -92,6 +92,23 @@ public async Task FilterAsync(ContentOptionsViewModel model, IQuery<ContentItem>
                 {
                     query.With<ContentItemIndex>(x => x.ContentType.IsIn(listableTypes.Select(t => t.Name).ToArray()));
                 }
+                else
+                {
+                    // Make sure we remove ContentItems from non-listable ContentTypes if
+                    // user doesn't have any unrestricted listable ContentType
+                    var nonListableTypes = new List<ContentTypeDefinition>();
+                    foreach (var ctd in _contentDefinitionManager.ListTypeDefinitions())
+                    {
+                        if (!ctd.GetSettings<ContentTypeSettings>().Listable)
+                        {
+                            nonListableTypes.Add(ctd);
```

Comment:
First, above line 91 we are doing

    if (listableTypes.Any())
    {
        query.With<ContentItemIndex>(x => x.ContentType.IsIn(listableTypes.Select(t => t.Name).ToArray()));
    }

Normally it should work without `if (listableTypes.Any())` so that we don't need this else block

If we still need to use the above `if (listableTypes.Any())`, if we go in this else block we already know that all content types are not listable, so maybe resolve once `.ListTypeDefinitions()` and hold it in a var , then above try to reduce the collection, otherwise use the whole collection (if any value).

But i think we should not use `if (listableTypes.Any())` line 91, and if it doesn't work we have to fix it on YesSql side. In yesSql, e.g. for `IsIn`, if there is no value it uses the following that is always false

                    if (values.Count == 0)
                    {
                        builder.Append(" 1 = 0");
                    }

Reply:
How would it know which Content Types are non-listable if we don't exclude them from that Query?
So here when we do IsIn() it adds on the Query : WHERE ContentType IN (1,2,3,4,5,6). But if we are a user that has access to none content type then the Query will not do WHERE ContentType IN ()... so it expects that we want all the ContentTypes that are in the ContentItemIndex table. Here, I don't understand how removing the if(listableTypes.Any()) condition would help at all. This listableTypes collection can have a count of 0 items even if you are an administrator on the website.

Reply:
Ok so it will do exactly the same but the Query will be bigger in the end.

Reply:
I'm sorry I don't see what you guys mean at all. 😄

Reply:
Oops, i missed the `_authorizationService.AuthorizeAsync()` thing ;)

Let me check it again

Reply:
I just did a code optimization to make it more obvious and readable.

Reply:
Okay, same kind of remarks but not on the whole content type collection but those for a given user ;)

About resolving once this collection, you did it somewhere in your last change by only using one loop. But here, if a type is not authorized for an user it would need to be added to the `nonListableTypes`, even it is listable.

But again, maybe i missed something else, the `IsIn` should work on an empty collection, so that if there is no content type that is authorized and listable, we just have to filter with a `IsIn` on an empty collection.

E.g. with the last change there is still a path where no filter is done at all.

But maybe i'm wrong, just as i understand the code

Good candidate for a skipe if you want ;)